orderer/common/cluster/clusterservice.go

```diff
@@ -0,0 +1,265 @@
+
+import (
+	"bytes"
+	"encoding/asn1"
+	"fmt"
+	"io"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric-protos-go/msp"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/util"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//go:generate mockery --dir . --name ClusterStepStream --case underscore --output ./mocks/
+
+// ClusterStepStream defines the gRPC stream for sending
+// transactions, and receiving corresponding responses
+type ClusterStepStream interface {
+	Send(response *orderer.ClusterNodeServiceStepResponse) error
+	Recv() (*orderer.ClusterNodeServiceStepRequest, error)
+	grpc.ServerStream
+}
+
+type AuthRequestVerifier interface {
+	VerifyAuthRequest(orderer.ClusterNodeService_StepServer, *orderer.ClusterNodeServiceStepRequest, TLSSessionBindingGetter, SignatureVerifier) (string, error)
+}
+
+type NodeIdentity struct {
+	ID       uint64
+	Identity []byte
+}
+
+// ClusterService defines the raft Service
+type ClusterService struct {
+	StreamCountReporter              *StreamCountReporter
+	Dispatcher                       Dispatcher
+	Logger                           *flogging.FabricLogger
+	StepLogger                       *flogging.FabricLogger
+	MinimumExpirationWarningInterval time.Duration
+	CertExpWarningThreshold          time.Duration
+	MembershipByChannel              map[string]map[uint64]*msp.SerializedIdentity
+	Lock                             sync.RWMutex
+	AuthVerifier                     AuthRequestVerifier
+}
+
+type AuthRequestSignature struct {
+	Version   int64
+	Timestamp string
+	FromId    string
+	ToId      string
+	Channel   string
+}
+
+// Step passes an implementation-specific message to another cluster member.
+func (s *ClusterService) Step(stream orderer.ClusterNodeService_StepServer) error {
+	s.StreamCountReporter.Increment()
+	defer s.StreamCountReporter.Decrement()
+
+	addr := util.ExtractRemoteAddress(stream.Context())
+	commonName := commonNameFromContext(stream.Context())
+	exp := s.initializeExpirationCheck(stream, addr, commonName)
+	s.Logger.Debugf("Connection from %s(%s)", commonName, addr)
+	defer s.Logger.Debugf("Closing connection from %s(%s)", commonName, addr)
+
+	// expect auth message is the first msg on the stream
+	request, err := stream.Recv()
+	if err == io.EOF {
+		s.Logger.Debugf("%s(%s) disconnected well before establishing the stream", commonName, addr)
+		return nil
+	}
+	if err != nil {
+		s.Logger.Warningf("Stream read from %s failed: %v", addr, err)
+		return err
+	}
+	channel, err := s.AuthVerifier.VerifyAuthRequest(stream, request, GetTLSSessionBinding, VerifySignature)
+	if err != nil {
+		s.Logger.Warnf("service authentication failed with error: %v", err)
+		return status.Errorf(codes.Unauthenticated, "access denied")
+	}
+
+	for {
```

Comment:
This pattern is vulnerable to impersonation attacks post configuration. 
Imagine that you have a node with ID of `13` and with a certificate `cert`. 
A reconfiguration is being done, and now node `13` has a certificate of `cert'`. 

The old stream that authenticated with certificate `cert` is still active, and someone can still send messages through that stream to impersonate `cert'`. 

What you need to do, is during re-configuration, you close all streams that have certificates that are no longer in use.